examples/common/pigweed/RpcService.cpp

```diff
@@ -0,0 +1,118 @@
+/*
+ *
+ *    Copyright (c) 2020 Project CHIP Authors
+ *    All rights reserved.
+ *
+ *    Licensed under the Apache License, Version 2.0 (the "License");
+ *    you may not use this file except in compliance with the License.
+ *    You may obtain a copy of the License at
+ *
+ *        http://www.apache.org/licenses/LICENSE-2.0
+ *
+ *    Unless required by applicable law or agreed to in writing, software
+ *    distributed under the License is distributed on an "AS IS" BASIS,
+ *    WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ *    See the License for the specific language governing permissions and
+ *    limitations under the License.
+ */
+
+#include "RpcService.h"
+
+#include <array>
+#include <span>
+#include <string_view>
+
+#include "pw_hdlc_lite/rpc_channel.h"
+#include "pw_hdlc_lite/rpc_packets.h"
+#include "pw_log/log.h"
+#include "pw_rpc/channel.h"
+#include "pw_status/status.h"
+#include "pw_stream/sys_io_stream.h"
+#include "pw_sys_io/sys_io.h"
+
+#include "logging/CHIPLogging.h"
+
+#include <array>
+
+namespace chip {
+namespace rpc {
+namespace {
+
+using std::byte;
+
+constexpr size_t kMaxTransmissionUnit = 1500;
+
+// Used to write HDLC data to pw::sys_io.
+pw::stream::SysIoWriter writer;
+
+::chip::rpc::Mutex * uart_mutex;
+
+template <size_t buffer_size>
+class ChipRpcChannelOutputBuffer : public pw::rpc::ChannelOutput
+{
+public:
+    constexpr ChipRpcChannelOutputBuffer(pw::stream::Writer & writer, uint8_t address, const char * channel_name) :
+        pw::rpc::ChannelOutput(channel_name), writer_(writer), address_(address)
+    {}
+
+    std::span<std::byte> AcquireBuffer() override
+    {
+        uart_mutex->Lock();
+        return buffer_;
+    }
+
+    pw::Status SendAndReleaseBuffer(std::span<const std::byte> buffer) override
+    {
+        PW_DASSERT(buffer.data() == buffer_.data());
+        if (buffer.empty())
+        {
+            uart_mutex->Unlock();
+            return pw::Status::Ok();
+        }
+        pw::Status ret = pw::hdlc_lite::WriteUIFrame(address_, buffer, writer_);
+        uart_mutex->Unlock();
+        return ret;
+    }
+
+private:
+    pw::stream::Writer & writer_;
```

Comment:
pedantic: I think CHIP usually follows the "mMember" convention for class member naming, but I don't know if anyone cares.

Reply:
done